be/queue.c

```diff
@@ -0,0 +1,448 @@
+
+	bq->bq_cfg = *cfg;
+	bq->bq_the_end = false;
+	bq->bq_enqueued = 0;
+	bq->bq_dequeued = 0;
+	M0_ALLOC_ARR(bq->bq_qitems, be_queue_qitems_nr(bq) *
+		     (sizeof(struct be_queue_item) + cfg->bqc_item_length));
+	M0_ALLOC_ARR(bq->bq_ops_put, bq->bq_cfg.bqc_producers_nr_max);
+	M0_ALLOC_ARR(bq->bq_ops_get, bq->bq_cfg.bqc_consumers_nr_max);
+	if (bq->bq_qitems == NULL ||
+	    bq->bq_ops_put == NULL ||
+	    bq->bq_ops_get == NULL) {
+		m0_free(bq->bq_ops_get);
+		m0_free(bq->bq_ops_put);
+		m0_free(bq->bq_qitems);
+		return M0_ERR(-ENOMEM);
+	}
+	m0_mutex_init(&bq->bq_lock);
+	bqop_tlist_init(&bq->bq_op_put_unused);
+	for (i = 0; i < bq->bq_cfg.bqc_producers_nr_max; ++i) {
+		bqop_tlink_init_at_tail(&bq->bq_ops_put[i],
+		                        &bq->bq_op_put_unused);
+	}
+	bqop_tlist_init(&bq->bq_op_put);
+	bqop_tlist_init(&bq->bq_op_get_unused);
+	for (i = 0; i < bq->bq_cfg.bqc_consumers_nr_max; ++i) {
+		bqop_tlink_init_at_tail(&bq->bq_ops_get[i],
+		                        &bq->bq_op_get_unused);
+	}
+	bqop_tlist_init(&bq->bq_op_get);
+	bqq_tlist_init(&bq->bq_q_unused);
+	for (i = 0; i < be_queue_qitems_nr(bq); ++i) {
+		bqq_tlink_init_at_tail(be_queue_qitem(bq, i),
+		                       &bq->bq_q_unused);
+	}
+	bqq_tlist_init(&bq->bq_q);
+	return M0_RC(0);
+}
+
+M0_INTERNAL void m0_be_queue_fini(struct m0_be_queue *bq)
+{
+	struct be_queue_wait_op *bwo;
+	struct be_queue_item    *bqi;
+	uint64_t                 i;
+
+	M0_ENTRY("bq="BEQ_F, BEQ_P(bq));
+	M0_ASSERT_INFO(bq->bq_enqueued == bq->bq_dequeued,
+	               "bq="BEQ_F, BEQ_P(bq));
+
+	m0_tl_for(bqq, &bq->bq_q, bqi) {
+		/*
+		 * M0_LOG() couldn't print the item buffer at once,
+		 * unfortunately. So let's just at least show the number of
+		 * items by printing every item.
+		 */
+		M0_LOG(M0_ERROR, "there is an item in the queue");
+	} m0_tl_endfor;
+	bqq_tlist_fini(&bq->bq_q);
+	m0_tl_for(bqop, &bq->bq_op_get, bwo) {
+		M0_LOG(M0_ERROR, "bq=%p bbo_data="BUF_F,
+		       bq, BUF_P(&bwo->bbo_data));
+	} m0_tl_endfor;
+	bqop_tlist_fini(&bq->bq_op_get);
+	for (i = 0; i < bq->bq_cfg.bqc_consumers_nr_max; ++i)
+		bqop_tlink_del_fini(&bq->bq_ops_get[i]);
+	bqop_tlist_fini(&bq->bq_op_get_unused);
+	/* if there was nothing in bq_q then the following list is empty */
+	bqop_tlist_fini(&bq->bq_op_put);
+	for (i = 0; i < bq->bq_cfg.bqc_producers_nr_max; ++i)
+		bqop_tlink_del_fini(&bq->bq_ops_put[i]);
+	bqop_tlist_fini(&bq->bq_op_put_unused);
+	for (i = 0; i < be_queue_qitems_nr(bq); ++i)
+		bqq_tlink_del_fini(be_queue_qitem(bq, i));
+	bqq_tlist_fini(&bq->bq_q_unused);
+	m0_mutex_fini(&bq->bq_lock);
+	m0_free(bq->bq_ops_put);
+	m0_free(bq->bq_ops_get);
+	m0_free(bq->bq_qitems);
+	M0_LEAVE();
+}
+
+M0_INTERNAL void m0_be_queue_lock(struct m0_be_queue *bq)
+{
+	m0_mutex_lock(&bq->bq_lock);
+}
+
+M0_INTERNAL void m0_be_queue_unlock(struct m0_be_queue *bq)
+{
+	m0_mutex_unlock(&bq->bq_lock);
+}
+
+static uint64_t be_queue_q_size(struct m0_be_queue *bq)
```

Comment:
the function name "be_queue_q_size" is ambiguous, maybe.
xxx_size means the capacity of the queue. It makes me to think it is the max number of items in this queue.
"be_queue_qitems_nr", on the contrary, makes me to think it is the current active number of items in the queue.

Reply:
Tried to make them less ambiguous. Please check.